insights/client/crypto.py

```diff
@@ -0,0 +1,270 @@
+        # directory. This is only supported since gnupg 2.1.18 (RHEL 8).
+        shutdown_process = subprocess.Popen(
+            ["/usr/bin/gpgconf", "--kill", "all"],
+            stdout=subprocess.PIPE, stderr=subprocess.PIPE,
+            universal_newlines=True,
+            env={"GNUPGHOME": self._home},
+        )
+        _, stderr = shutdown_process.communicate()
+        if shutdown_process.returncode != 0:
+            stderr = "\n".join("stderr: {line}".format(line=line) for line in stderr.split("\n") if len(line))
+            logger.debug(
+                "could not kill the GPG agent, got return code {rc}: \n{err}".format(
+                    rc=shutdown_process.returncode, err=stderr
+                )
+            )
+
+    def _cleanup(self):
+        """Clean up GPG environment."""
+        if self._supports_cleanup_socket():
+            self._cleanup_socket()
+
+        # Older systems do not support `gpgconf --kill`.
+        # The socket may remove its socket file after `rmtree()` has determined
+        # it should be deleted, but before the actual deletion occurs.
+        # This would cause a FileNotFoundError/OSError.
+        for _ in range(5):
+            try:
+                shutil.rmtree(self._home)
+                break
+            except OSError as exc:
+                if exc.errno == errno.ENOENT:
+                    # The file has already been removed by the `gpg-agent`.
+                    continue
+                raise
+        else:
+            logger.debug(
+                "could not clean up temporary GPG directory '{path}'".format(path=self._home)
+            )
+
+    def _run(self, command):
+        """Run the actual command.
+
+        :returns: The result of the shell command.
+        :rtype: GPGCommandResult
+        """
+        self._raw_command = ["/usr/bin/gpg", "--homedir", self._home] + command
+        process = subprocess.Popen(
+            self._raw_command,
+            stdout=subprocess.PIPE, stderr=subprocess.PIPE,
+        )
+        stdout, stderr = process.communicate()
+
+        result = GPGCommandResult(
+            ok=process.returncode == 0,
+            return_code=process.returncode,
+            stdout=stdout.decode("utf-8"),
+            stderr=stderr.decode("utf-8"),
+            command=self,
+        )
+
+        if result.ok:
+            logger.debug("gpg command {command}: ok".format(command=command))
+        else:
+            logger.debug(
+                "gpg command {command} returned non-zero code: {result}".format(
+                    command=command, result=result,
+                )
+            )
+
+        return result
+
+    def evaluate(self):
+        """Run the command.
+
+        :returns: The result of the shell command.
+        :rtype: GPGCommandResult
+        """
+        try:
+            result = self._setup()
+            if not result.ok:
+                logger.debug("gpg setup failed")
```

Comment:
perhaps I'd bump this log message to warning, as it seems a severe enough failure

Reply:
When we bump the log message level in `_setup()`, we don't need to do it here again. We should either leave it as-is, or drop it.

Reply:
either is fine -- my point is that kind of "important failures" ought to be shown somewhere else than the log file only